Lecture4

bulters opened the pull request:
### Solutions for Lecture 4

Added partial performance improvements, that did not yield desired results yet.

cc @chshersh @vrom911

bulters committed:
Lecture 3 exercises

bulters committed:
Cleanup Lecture 3 solutions

bulters committed:
Eta reduce toShortString

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

bulters committed:
Replace Gold 0 with mempty

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

bulters committed:
Replace second List concat by cons in List1 <>

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

bulters committed:
Make appendDiff3 implementation dump but without constraint

bulters committed:
Simplify foldMap for List1

bulters committed:
Merge branch 'lecture2' of github.com:bulters/exercises into lecture3

bulters committed:
Don't cheat in the apply implementation

bulters committed:
Don't need to import Data.List anymore

Since the use of `nub` is now eliminated.

bulters committed:
Do mathy version of daysTo

Left the old implementation as comment, given it's still
nice ;-)

bulters committed:
Add comments to indicate where imports can be added (#53)

* Add comments to indicate where imports can be added

* Add arrows to indicate where imports can be added

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

fetsh committed:
Replace import comments syntax (#57)

chshersh committed:
Fix huge file generator (#58)

chshersh committed:
Documentation and test improvements to the course (#59)

fetsh committed:
Prevent generated big.csv being commmited (#60)

chshersh committed:
Lecture 4 clarifications and suggestions (#62)

Flinner committed:
Add tests to `Semigroup`s (#64)

* feat: check `List1`'s `Semigroup`

* feat: check `MaxLen`'s `Semigroup`

bulters committed:
First dump implementation for Lecture 4 exercises

bulters committed:
Add some performance optimizations

bulters committed:
Merge remote-tracking branch 'origin/lecture3' into lecture4

bulters committed:
Cleanup spacing around $

chshersh commented on src/Lecture4.hs:
```diff
@@ -223,7 +306,9 @@ implement the next task.
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
+combineRows (r :| [])      = rowToStats r
+combineRows (r :| [r'])    = rowToStats r <> rowToStats r'
+combineRows (r :| (r':rs)) = rowToStats r <> combineRows (r' :| rs)
```
Without providing detailed feedback yet and taking only a quick look, this function looks like the main problem. This function is not tail-recursive and hence leaks memory.

To understand the issue here better, consider the following implementation of `sum`:

```haskell
sum :: [Int] -> Int
sum [ ] = 0
sum (x : xs) = x + sum xs
```

If you try to evaluate sum `[1 .. 1000]` you will get the following result, according to the implementation:

```haskell
sum [1 .. 1000]
= 1 + sum [2 .. 1000]
= 1 + (2 + sum [3 .. 1000])
= 1 + (2 + (3 + sum [4 .. 1000]))
...
```

and so on. This `sum` function doesn't add numbers immediately. It stores evaluation results in memory until the end of the list. The same is happening with the `combineRows` function. 

This has nothing to do with lazy evaluation, it's just a general problem with not tail-recursive functions in every language 🙂 

To fix this problem, you need to either implement a recursive function like `go` with the bang-patterned accumulator `!acc` or you can implement a separate function that combines non-empty list of `Stats`:

```haskell
combineRows :: NonEmpty Stats -> Stats
```

This function can be easily implemented in one line using fold. And if you use `foldl'` specifically, it will have the desired performance 🙂 

And then, using this function, you can implement `combineRows` in a single line as well 🏎️

bulters committed:
Refactored combineRows into TCO variant.

Performance is now acceptable. Timewise large run in 24 minutes,
space wise, memory use is constant around 200 megabytes.

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +141,34 @@ errors. We will simply return an optional result here.
+                  in case mp of
+                    Just "" -> Nothing
+                    Just x  -> Just x
+                    Nothing -> Nothing
```
I think that conversion through `readMaybe` and `show` is a bit redundant here 🤔 
In the task, the only required validation for for product names is being a non-empty string. So you can write a "simpler" parser here 🙂 

```suggestion
parseProduct :: String -> Maybe String
parseProduct ps
   | null ps = Nothing
   | otherwise = Just ps
```

bulters replied:
This one has been refactored away by the use of `MonadFail`/`guard`

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +141,34 @@ errors. We will simply return an optional result here.
+parseTradeType = readMaybe
+
+parseCosts :: String -> Maybe Int
+parseCosts cs = let mc = readMaybe cs
+                in case mc of
+                  Just x  -> if x < 0 then Nothing else Just x
+                  Nothing -> Nothing
```
That looks great 👏🏻 
Alternatively, you can use the `Monad` instance for `Maybe` here to write a bit shorter:

```suggestion
parseCosts cs =
    readMaybe cs >>= \x -> if x < 0 then Nothing else Just x
```

bulters replied:
This one has been refactored away by the use of `MonadFail`/`guard`

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +141,34 @@ errors. We will simply return an optional result here.
+                 tt <- parseTradeType ttp
+                 c <- parseCosts cp
+                 return Row { rowProduct = p, rowTradeType = tt, rowCost = c }
+               _ -> Nothing
```
You have an excellent monadic implementation 🔝 
And I see that you're even using `do`-notation for `Maybe` here 😎 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow s = do
    [p, ttp, cp] <- Just $ splitOn "," s
    guard $ p /= ""
    tt <- readMaybe rawTrade
    c <- readMaybe cp
    guard $ cost >= 0
    return Row { rowProduct = p, rowTradeType = tt, rowCost = c }
```

chshersh commented on src/Lecture4.hs:
```diff
@@ -180,7 +219,24 @@ The 'Stats' data type has multiple fields. All these fields have
 
+-- <!> defines a strict version of <> for the Maybe type.
+(<!>) :: Semigroup a => Maybe a -> Maybe a -> Maybe a
+Nothing   <!> Nothing   = Nothing
+(Just !x) <!> Nothing   = Just x
+Nothing   <!> (Just !y) = Just y
+(Just !x) <!> (Just !y) = Just (x <> y)
```
That's absolutely correct implementation 👍🏻 
You can write it a bit shorter though by pattern matching only on the final result:

```suggestion
a <!> b = case a <> b of
    Nothing -> Nothing
    Just !x -> x
```

It's also a bit lazier but in a good way 👼🏻

chshersh commented on src/Lecture4.hs:
```diff
@@ -196,8 +252,35 @@ row in the file.
+
 rowToStats :: Row -> Stats
-rowToStats = error "TODO"
+rowToStats r
+  | rowTradeType r == Buy = withBuy (baseStats r) r
+  | rowTradeType r == Sell = withSell (baseStats r) r
+  | otherwise = Stats 0 0 (Max 0) (Min 0) Nothing Nothing Nothing Nothing (MaxLen "")
```
Since `TradeType` can be only `Buy` or `Sell`, pattern matching works better here 🙂 

```suggestion
rowToStats r = case rowTradeType r of
    Buy  -> withBuy (baseStats r) r
    Sell -> withSell (baseStats r) r
```

It also will be better because if in the future you decide to add more constructors to the `TradeType`, GHC  will produce a compile-time warning to remind you to update this function. With the current implementation you won't get a warning.

chshersh commented on src/Lecture4.hs:
```diff
@@ -254,7 +353,7 @@ the file doesn't have any products.
 
 calculateStats :: String -> String
-calculateStats = error "TODO"
+calculateStats ls = displayStats . combineRows . fromList . mapMaybe parseRow $ lines ls
```
`fromList` is an unsafe function. For example, after `mapMaybe` you may get an empty list and `fromList` will throw an error

chshersh reviewed (approved):
:tada: Fantastic work! Congratulations on finishing all **four** lectures! 🥳 
You're truly a Functional Adept now 🧠 
I hope you had fun :man_dancing: :microphone:

If you would like to receive a pretty certificate about the course completion, please, send your full name and the link to this PR to `haskell.beginners2022@gmail.com` 📧 

If you have time, I would appreciate your feedback on the course:

* [Haskell Beginners 2022 Course Feedback](https://docs.google.com/forms/d/e/1FAIpQLSfCkST64dnDCYkz5mezkVyLfWjeskSxhV0Y_cYxUaJw0U9ZhQ/viewform)

If you like my work, you can [sponsor me on GitHub](https://github.com/sponsors/chshersh) 💖

You can also ask any questions here or anywhere else and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

<hr>

If you would like to continue your journey to Haskell, check out the [Haskell Knowledge Map](https://kowainik.github.io/images/Haskell_Knowledge_Map.png) for the list of topics to explore 🗺️

bulters committed:
Use null to parse empty or Nothing product strings

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

bulters committed:
Refactor parseProduct to use MonadFail and guard

bulters committed:
Refactor rowToStats to pattern match on trade type

this allows potential future addition of other types to throw
a compiler error when not catching them properly.

bulters committed:
Refactor <!> to make it more lazy

... and easier.

bulters committed:
Refactor calculateStats to deal with faulty input

bulters committed:
Refactor to use Lazy ByteString instead of String

This reduces runtime from ~20 minutes to ~30 seconds.

chshersh commented on exercises.cabal:
```diff
@@ -69,12 +69,17 @@ library
+  build-depends:       base
+                     , split
+		     , bytestring
+		     , text
```
That's a very easy way to get performance but you can compiler your code with better optimizations 🙂 
```suggestion
		     , text
  ghc-options: -O2
```

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +143,15 @@ errors. We will simply return an optional result here.
+
+parseRow :: BL.ByteString -> Maybe Row
+parseRow s = do
+  [pp, ttp, cp] <- Just $ BLC.split ',' s
```
It would be more efficient to use `split` that takes `Word8` instead of `Char`

* https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/Data-ByteString-Lazy.html#v:split

In fact, it might be even better to pattern match on the result of `lines` for lazy `ByteString` and convert a list of lazy bytestrings to the list of strict bytestrings. Lazy bytestring is basically just a linked list of strict bytestrings and strict bytestring is an array of bytes.

When you split lazy bytestring using `lines`, semantically, you get a list of lists of lazy bytestrings. Since in the huge file every single line is very short, each lazy bytestring will be just a singleton list. So you can write a helper function to normalize your data after lines and parse strict bytestring here which will be much more efficient 🙂 

The code won't be that beautiful anymore 😅 But if you want to squeeze max performance in Haskell, sometimes you need to go on a lower level...

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +143,15 @@ errors. We will simply return an optional result here.
+parseRow s = do
+  [pp, ttp, cp] <- Just $ BLC.split ',' s
+  guard $ pp /= ""
+  tt <- readMaybe $ BLC.unpack ttp
```
Going through `String` and using `readyMaybe` is quite inefficient because `String` and `Read`-parsing is very slow. For the `TradeType` enum, you can drop spaces and pattern-match on the prefix of the remaining string for the most optimal performance.

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +143,15 @@ errors. We will simply return an optional result here.
+  [pp, ttp, cp] <- Just $ BLC.split ',' s
+  guard $ pp /= ""
+  tt <- readMaybe $ BLC.unpack ttp
+  c <- readMaybe $ BLC.unpack cp
```
The same applies here. I can't suggest the most efficient way to parse `Int` from `ByteString`, didn't have this task before 🤔 

Asked on Twitter, maybe someone will reply 😅

chshersh replied:
And here is the answer:

https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/Data-ByteString-Char8.html#v:readInt

Twitter helps 🤣

chshersh commented on src/Lecture4.hs:
```diff
@@ -131,8 +143,15 @@ errors. We will simply return an optional result here.
+  tt <- readMaybe $ BLC.unpack ttp
+  c <- readMaybe $ BLC.unpack cp
+  guard $ c >= 0
+  return Row {rowProduct = BLC.unpack pp, rowTradeType = tt, rowCost = c }
```
If you're working with `ByteString`, you can store `ByteString` directly in `Row` and in `MaxLen`. No need to have extra conversions here and spend more memory than needed.

Decreasing memory footprint in Haskell actually increases performances because GC needs to do less work.

chshersh commented on src/Lecture4.hs:
```diff
@@ -154,7 +173,10 @@ string.
 -}
 instance Semigroup MaxLen where
-
+  a <> b = case compare (length (unMaxLen a)) (length (unMaxLen b)) of
```
If `MaxLen` will store `ByteString` instead of `String`, the `length` function will run in `O(1)` time instead of `O(n)`. Also, you can use `>=` operator here to reduce pattern matching to a single primitive operation 🙂

chshersh reviewed (approved):
@bulters I've added some more notes on what you can do to increase performance 🙂 

Just in case you want to continue playing benchmarking game further 😅
